ko/pl.rst

```diff
@@ -0,0 +1,1383 @@
+    <inner_function_decl> ::=
+        FUNCTION <identifier> [ ( <seq_of_parameters> ) ] RETURN <type_spec> { IS | AS } [ <seq_of_declare_specs> ] <body> ;
+
+    <seq_of_parameters> ::= <parameter> [, <parameter> ...]
+    <parameter> ::= <identifier> [ { IN | IN OUT | OUT } ] <type_spec>
+    <body> ::= BEGIN <seq_of_statements> [ EXCEPTION <seq_of_handlers> ] END [ <label_name> ]
+    <seq_of_declare_specs> ::= <declare_spec> [ <declare_spec> ... ]
+    <seq_of_statements> ::= <statement> ; [ <statement> ; ... ]
+    <seq_of_handlers> ::= <handler> [ <handler> ... ]
+    <handler> ::= WHEN <exception_name> [ OR <exeption_name> OR ... ] THEN <seq_of_statements>
+    <exception_name> ::= OTHER | identifier
+
+* *parameter*: 파라메터는 IN, IN OUT, OUT 세 가지 경우로 선언할 수 있다.
+* *type_spec*: :ref:`데이터 타입 <types>` 절에서 설명한 시스템 제공 타입
+* *body*: 필수적으로 하나 이상의 실행문과 선택적으로 몇 개의 Exception 핸들러로 구성된다.
+* *declare_spec*: 변수, 상수, Exception, 커서, Autonomous Transaction, 내부 프로시저, 내부 함수 선언 중 하나
+* *statement*: 아래 실행문 절 참조
+* *handler*: OR로 연결된 하나 이상의 Exception 이름들에 대하여 실행할 실행문들을 지정한다.
+* *exception_name*: OTHER인 경우 아직까지 매치되지 않은 모든 Exception에 매치된다. 아닌 경우는 시스템 Exception이거나 사용자 정의 Exception을 나타낸다.
+
+함수 *body*\에서는 RETURN 절에 지정된 타입에 맞는 값을 반환해야 한다.
+함수가 *body* 끝에 도달할 때까지 RETURN 문을 만나지 못하면 에러가 발생한다.
+프로시저의 RETURN 문에 반환값을 지정하면 에러이다.
+
+(내부) 프로시저/함수를 선언하면 자기 자신을 실행부에서 참조할 수 있다. 즉, 재귀호출이 가능하다.
+
+.. code-block:: sql
+
+    CREATE FUNCTION choose(m INT, n INT) RETURN INT AS
+
+        invalid_argument EXCEPTION;
+
+        -- 내부 함수 선언
+        FUNCTION factorial(n INT) RETURN INT AS
+        BEGIN
+            IF n < 0 THEN
+                RAISE invalid_argument;
+            ELSIF n <= 1 THEN
+                RETURN 1;
+            ELSE
+                RETURN n * factorial(n - 1);    -- 재귀 호출
+            end if;
+        END;
+    BEGIN
+        IF n > m OR n < 0 THEN
+            RAISE invalid_argument;
+        ELSE
+            RETURN factorial(m) / factorial(n) / factorial(m - n);
+        END IF;
+    END;
+
+동일한 선언부에서 선언된 내부 프로시저/함수끼리는 상호 재귀호출도 가능하다.
+
+.. code-block:: sql
+
+    CREATE PROCEDURE ping_pong AS
+
+        PROCEDURE ping AS
+        BEGIN
+            put_line('ping ->');
+            pong();     -- 상호 재귀 호출
+        END;
+
+        PROCEDURE pong AS
+        BEGIN
+            put_line('      <- pong');
+            ping();     -- 상호 재귀 호출
+        END;
+    BEGIN
+        ping();
+    END;
+
+.. _auto_tran:
+
+Autonomous Transaction 선언
+===========================
+::
+
+    <autonomous_transaction_decl> ::=
+        PRAGMA AUTONOMOUS_TRANSACTION ;
+
+이 선언문을 포함한 스토어드 프로시저/함수는 호출한 쪽의 트랜잭션에 포함되는 것이 아니라,
+독자적인 자신만의 트랜잭션 안에서 실행된다.
+이 경우에 스토어드 프로시저/함수 실행 도중에 DB 변경이 있었는데도 COMMIT이나 ROLLBACK이 실행되지 않으면 에러이다.
+그리고, COMMIT이나 ROLLBACK을 실행해도 호출한 쪽의 트랜잭션의 진행에는 영향을 미치지 않는다.
+
+Autonomous Transaction으로 선언되지 않은 스토어드 프로시저/함수는 호출한 쪽의 트랜잭션 안에 포함된다.
+이 경우에는 스토어드 프로시저/함수 안에서 호출한 COMMIT이나 ROLLBACK이 아무 동작도 하지 않고 무시된다.
```

Comment:
How about adding "Changes in the stored procedure/function are reflected only when the caller commits them"?

Reply:
Is it okay to ignore it? I've said previous meeting that 'Autonomous Transaction' and 'commit rollback' should be considered separately. I remember that there was no particular objection at the meeting, and I wonder why such a conclusion was reached.

Reply:
I remember the decision was "follow the current JSP semantics, that is, ignoring them" for the procedures and functions without AUTONOMOUS TRANSACTION specification.
We need to talk about this again.